fix static viz waterfall colors

alxnddr opened the pull request:
Closes https://github.com/metabase/metabase/issues/20621

## Changes

Fixes default colors of static waterfall charts. Static viz JS code expects the instance colors from the BE to generate all necessary shades. If the instance colors are empty it uses the default ones.

We plan to remove the following chart types soon so I did not remove hardcoded colors from their `const layout = ...` definition.

```
CATEGORICAL_AREA_CHART_TYPE
CATEGORICAL_BAR_CHART_TYPE
CATEGORICAL_LINE_CHART_TYPE
TIME_SERIES_AREA_CHART_TYPE
TIME_SERIES_BAR_CHART_TYPE
TIME_SERIES_LINE_CHART_TYPE
```

<img width="1439" alt="Screen Shot 2022-08-17 at 11 40 36 PM" src="https://user-images.githubusercontent.com/14301985/185228077-a0d7e146-5989-4883-b806-a14fb9f5dc70.png">

<img width="633" alt="Screen Shot 2022-08-17 at 11 40 09 PM" src="https://user-images.githubusercontent.com/14301985/185228014-2496ceb5-8236-4ff6-8838-839534825495.png">

alxnddr committed:
fix static viz waterfall colors

adam-james-v commented on src/metabase/pulse/render/body.clj:
```diff
@@ -776,9 +776,9 @@
-                                   :waterfallTotal (or (:waterfall.total_color viz-settings) (nth colors 0))
-                                   :waterfallPositive (or (:waterfall.increase_color viz-settings) (nth colors 1))
-                                   :waterfallNegative (or (:waterfall.decrease_color viz-settings) (nth colors 2)))
+                                   :waterfallTotal (:waterfall.total_color viz-settings)
```
This works, but just be aware that it will explicitly set `:waterfallTotal` to `nil` if there is no `:waterfall.total_color` in viz-settings. So, as long as the JS will properly handle a key existing but its value being `nil`, then you're good to go.

adam-james-v commented on src/metabase/pulse/render/js_svg.clj:
```diff
@@ -115,7 +116,8 @@
                                                   (map (fn [[k v]] [(name k) v]) labels)
-                                                  (json/generate-string settings)))]
+                                                  (json/generate-string settings)
+                                                  (json/generate-string (public-settings/application-colors))))]
```
Looks good. Keep in mind that `public-settings/application-colors` may be an empty map, and if it has values, it _only_ has values which are different than defaults. For instance, if only the 'brand' color was changed to `#00FF00` in the admin panel, then this map will look like: `{:brand "#00FF00"}`, with no other colors set.

I think the changes you've made in this PR will properly get colors anyway, just wanted to make that clear about the setting coming from the backend.

adam-james-v reviewed (approved):
This looks nice! I like that you're removing hardcoded color values. This will also make my [color PR](https://github.com/metabase/metabase/pull/24834) AND my [waterfall colors PR](https://github.com/metabase/metabase/pull/24800) obsolete, right?

alxnddr committed:
tests

alxnddr commented:
@adam-james-v yes, seems like this change fixes the colors so you can close the other PRs. Thanks for bringing up the issue with defaults which led us to the appropriate solution!

adam-james-v committed:
Pass colors into StaticChart as third argument

This might not be the best design, so changes are welcome, but I added this to properly consider the case where the
application-colors are different, but the chart has no viz-settings: application-colors were being ignored (as they
were passed into options, but not pulled out of options into the colors arg).

WiNloSt commented on frontend/src/metabase/static-viz/components/CategoricalWaterfallChart/CategoricalWaterfallChart.jsx:
```diff
@@ -74,13 +68,14 @@ const CategoricalWaterfallChart = ({
     settings?.showTotal,
   );
-  const colors = settings?.colors;
+  const colors = settings?.colors ?? {};
```
Does this change imply `settings?.colors` could be `null`? Since `colors` will be passed to

https://github.com/metabase/metabase/blob/5e9f03ae4b8bd40cae09ab375b5715d3b5b9ce82/frontend/src/metabase/static-viz/lib/colors.ts#L18-L26

as the first argument which already has a default value set. The only reason `?? {}` would be needed is when passing `null` which should crash `getWaterfallColors`.

If that's the case I think it's better to not let people assume things and type `settings.colors` as nullable.

WiNloSt commented on resources/frontend_shared/static_viz_interface.js:
```diff
@@ -47,11 +47,12 @@ function combo_chart(series, settings, colors) {
 }
 
-function timeseries_waterfall(data, labels, settings) {
+function timeseries_waterfall(data, labels, settings, colors) {
```
Do you think this should better name `instanceColor` as in
https://github.com/metabase/metabase/blob/5e9f03ae4b8bd40cae09ab375b5715d3b5b9ce82/frontend/src/metabase/static-viz/lib/colors.ts#L4

At glance, I might thought this is some kind of arbitrary color setting for this specific chart rather than an instance color.

Also, I think the `colors` type for `combo-chart` is out of date.
https://github.com/metabase/metabase/blob/5e9f03ae4b8bd40cae09ab375b5715d3b5b9ce82/frontend/src/metabase/static-viz/components/LineAreaBarChart/LineAreaBarChart.tsx#L21

https://github.com/metabase/metabase/blob/5e9f03ae4b8bd40cae09ab375b5715d3b5b9ce82/frontend/src/metabase/static-viz/components/LineAreaBarChart/types.ts#L1-L6

I can clearly see that we pass instance colors to combo chart 

https://github.com/metabase/metabase/blob/5e9f03ae4b8bd40cae09ab375b5715d3b5b9ce82/src/metabase/pulse/render/body.clj#L318

`settings` is a value from calling `->ts-viz` which is then passed to `js-svg/combo-chart` on line 511

https://github.com/metabase/metabase/blob/5e9f03ae4b8bd40cae09ab375b5715d3b5b9ce82/src/metabase/pulse/render/body.clj#L506-L511

alxnddr replied:
Thanks, renamed that. Regarding combo-chart, it expects colors in a different way because that was before the instance colors settings revamp. I'll update it separately

WiNloSt commented on frontend/src/metabase/static-viz/index.js:
```diff
@@ -4,6 +4,6 @@ import StaticChart from "./containers/StaticChart";
 export function RenderChart(type, options) {
   return ReactDOMServer.renderToStaticMarkup(
-    <StaticChart type={type} options={options} />,
+    <StaticChart type={type} options={options} colors={options.colors} />,
```
Not sure if we need to pass `colors` as an extra prop since it's already in `StaticChart`. I guess you do that since you want to assign a default value to `colors` but I don't think you actually need to do that since both `null` and `undefined` can be destructured into an object just fine.

- https://github.com/metabase/metabase/blob/5e9f03ae4b8bd40cae09ab375b5715d3b5b9ce82/frontend/src/metabase/static-viz/containers/StaticChart/StaticChart.tsx#L29-L30
- https://github.com/metabase/metabase/blob/5e9f03ae4b8bd40cae09ab375b5715d3b5b9ce82/frontend/src/metabase/static-viz/lib/colors.ts#L4-L5
- ![image](https://user-images.githubusercontent.com/1937582/185309714-81733da3-fd79-4081-b4fd-afb084b66c7c.png)

So, if I'm not mistaken you should be fine dropping default values for both colors props and even remove the `colors` prop from `StaticChart`

WiNloSt reviewed (approved):
Sorry, I haven't tested this, but the code looks fairly straightforward. I also left some comments for some clarifications.

And the Percy test failed. Which is expected since we're fixing the color on the waterfall chart, but just let you know in case you don't see that.

alxnddr committed:
review
